### PR TITLE
Restructure certificate redeploy playbooks

### PR DESCRIPTION
**Pull request by abutcher**
Restructure certificate redeploy playbooks to support redeploying certificates for control plane and nodes separately to give admins more control over certificate redeploy. This mainly allows node certificate redeploy to be ran individually against sets of nodes or nodes which may have failed as a result from original certificate redeployment.

~~These changes also replace the currently separate etcd CA certificate with the OpenShift CA following a redeploy and as the default for new clusters.~~

### OpenShift CA

Redeploys the OpenShift CA certificate by generating a new CA certificate and distributing an updated bundle to all components including client kubeconfigs and node CA trust.

Serial restarts: etcd, master services, node service and docker.

```
ansible-playbook -i inventory playbooks/byo/openshift-cluster/redeploy-openshift-ca.yml
```

### OpenShift Certificates

Redeploys etcd, master, node, router and registry certificates. This playbook does not redeploy the OpenShift CA certificate and uses the current CA certificate to sign new certificates.

Serial restarts: etcd, master services, and node service.

```
ansible-playbook -i inventory playbooks/byo/openshift-cluster/redeploy-certificates.yml
```

### Etcd Certificates

This playbook only redeploys etcd certificates including master client certificates.

Serial restarts: etcd and master services.

```
ansible-playbook -i inventory playbooks/byo/openshift-cluster/redeploy-etcd-certificates.yml
```

### Master Certificates

This playbook only redeploys master certificates.

Serial restarts: master services.

```
ansible-playbook -i inventory playbooks/byo/openshift-cluster/redeploy-master-certificates.yml
```

### Node Certificates

This playbook only redeploys node certificates.

Serial restarts: node services.

```
ansible-playbook -i inventory playbooks/byo/openshift-cluster/redeploy-node-certificates.yml
```

### Registry and Router Certificates

These playbooks replace installer created certificates for the router and registry. If custom certificates are in use for these components they must be replaced manually.

```
ansible-playbook -i inventory playbooks/byo/openshift-cluster/redeploy-router-certificates.yml
ansible-playbook -i inventory playbooks/byo/openshift-cluster/redeploy-registry-certificates.yml
```

### CODE REVIEW

File: playbooks/common/openshift-cluster/redeploy-certificates/nodes-evacuation.yml [hunk 1]
```diff
@@ -0,0 +1,62 @@
+---
```
Comment (sferich888) [inline playbooks/common/openshift-cluster/redeploy-certificates/nodes-evacuation.yml]: We should put a comment in this script that explains why this is done.

File: playbooks/common/openshift-cluster/redeploy-certificates/redeploy-certificates.yml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+---
+- include: control-plane.yml
+
```
Comment (sferich888) [inline playbooks/common/openshift-cluster/redeploy-certificates/redeploy-certificates.yml]: We probably want to have a master/control-plane restart (option as well). The thought here, is to provide a way to just restart the control plane if there is an issue (with the general playbook).

File: playbooks/common/openshift-cluster/redeploy-certificates/redeploy-certificates.yml [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+- include: nodes.yml
```
Comment (sferich888) [inline playbooks/common/openshift-cluster/redeploy-certificates/redeploy-certificates.yml]: We may want to rename this to something like re-certificate-node.yaml

**Comment by detiber** [comment]
@abutcher haven't done a full review, but I'm thinking we'd want to try and do things in the following fashion:

- etcd certs
  - if etcd allows for multiple certs in the configured CA file:
    - regenerate etcd CA and add new cert to existing CA file
    - for each etcd host:
      - regenerate etcd service/peer certs
      - restart etcd
    - for each master host:
      - regenerate etcd client certs
      - restart api services
  - if etcd does not allow for multiple certs in the configured CA file:
    - regenerate etcd CA and add new cert to existing CA file
    - regenerate all etcd service/peer certs
    - regenerate all etcd client certs
    - restart etcd (this will start an outage of the api service until the master api service restarts have completed).
    - restart master api service

- regenerate OpenShift CA and add newly generated CA crt to the CA file
  - I believe this was supported for 3.2, but may require 3.3. I don't believe 3.1 would support this.
- for each master:
  - regenerate openshift certificates
  - restart api/controllers services
  - regenerate cli certificates
- for each node:
  - regenerate openshift certificates
  - restart node service

File: playbooks/byo/openshift-cluster/redeploy-certificates.yml [hunk 4]
```diff
@@ -14,7 +14,36 @@
   gather_facts: no
   tasks:
   - include_vars: ../../byo/openshift-cluster/cluster_hosts.yml
+  - set_fact:
+      openshift_deployment_type: "{{ deployment_type }}"
```
Comment (detiber) [inline playbooks/byo/openshift-cluster/redeploy-certificates.yml]: Any particular reason to use `set_fact` here instead of `vars`? vars has a lower runtime performance impact.

File: playbooks/byo/openshift-cluster/redeploy-certificates.yml [hunk 5]
```diff
@@ -14,7 +14,36 @@
...
+  tags:
+  - always
+
+- include: ../../common/openshift-cluster/initialize_facts.yml
+  tags:
+  - always
+
+- include: ../../common/openshift-cluster/initialize_openshift_version.yml
+  tags:
+  - always
```
Comment (detiber) [inline playbooks/byo/openshift-cluster/redeploy-certificates.yml]: It looks like everything from line 2 to here is boilerplate that is common to all of the playbooks, if it's being included in each of the included playbooks, it shouldn't be needed here. I'm also wondering if it would make sense to have a playbook that encapsulates these common steps to avoid the boilerplate feel.

File: playbooks/common/openshift-cluster/redeploy-certificates/nodes-evacuation.yml (hunk 1 above)
Comment (detiber) [inline playbooks/common/openshift-cluster/redeploy-certificates/nodes-evacuation.yml]: Node evacuation, service restarts, etc all seem to be more generic actions that aren't necessarily tied to certificate redeployment. Should they exist outside of the redeploy-certificates dir and be included in from more generic directories?

File: playbooks/byo/openshift-cluster/redeploy-certificates.yml (hunk 4 above)
Comment (abutcher) [inline playbooks/byo/openshift-cluster/redeploy-certificates.yml]: None other than setting it once rather than in vars on every include.

File: playbooks/common/openshift-cluster/redeploy-certificates/nodes-evacuation.yml (hunk 1 above)
Comment (abutcher) [inline playbooks/common/openshift-cluster/redeploy-certificates/nodes-evacuation.yml]: +1 I'll separate them.

File: playbooks/byo/openshift-cluster/redeploy-certificates.yml (hunk 4 above)
Comment (detiber) [inline playbooks/byo/openshift-cluster/redeploy-certificates.yml]: Ah, that's right...  We probably need to start exposing openshift_deployment_type in the inventory, and just mapping deployment_type to openshift_deployment_type for legacy use.

File: library/kubeclient_ca.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        return module.exit_json(changed=(len(changes) > 0))
+
+    # ignore broad-except error to avoid stack trace to ansible user
+    # pylint: disable=broad-except
+    except Exception as error:
+        return module.fail_json(msg=str(error))
+
+
+# pylint: disable=wrong-import-position
+from ansible.module_utils.basic import AnsibleModule  # noqa: F402
```
Comment (detiber) [inline library/kubeclient_ca.py]: I think this can be moved up to the top now that we are using Ansible > 2.1

File: roles/etcd_client_certificates/tasks/main.yml [hunk 7]
```diff
@@ -76,7 +76,7 @@
     src: "{{ etcd_ca_cert }}"
     dest: "{{ etcd_generated_certs_dir}}/{{ etcd_cert_subdir }}/{{ etcd_cert_prefix }}ca.crt"
     state: hard
-  when: etcd_client_certs_missing | bool
+  when: etcd_server_certs_missing | bool
```
Comment (detiber) [inline roles/etcd_client_certificates/tasks/main.yml]: Was this change intentional?

File: roles/openshift_ca/tasks/main.yml [hunk 8]
```diff
@@ -40,13 +40,18 @@
...
-    master_ca_missing: "{{ true if openshift_certificates_redeploy | default(false) | bool
-                           else False in (g_master_ca_stat_result.results
-                                         | oo_collect(attribute='stat.exists')
-                                         | list) }}"
+    master_ca_missing: "{{ False in (g_master_ca_stat_result.results
+                                     | oo_collect(attribute='stat.exists')
+                                     | list) }}"
   run_once: true
 
+- debug: var=master_ca_missing
```
Comment (detiber) [inline roles/openshift_ca/tasks/main.yml]: probably want to remove this before merging.

**Comment by abutcher** [comment]
aos-ci-test

**Comment by DanyC97** [comment]
@abutcher am i right to believe this wokr will be available in Origin 1.5 ?

**Comment by abutcher** [comment]
Correct.
